### PR TITLE
POEM 044 Implementation: Warning Refactor

### LINKED ISSUE #1944: POEM 044 Implementation: Warnings refactor
### Summary of Issue

The proposed refactor of OpenMDAO warnings has been added as POEM [044](OpenMDAO/POEMs#100)

### Issue Type

- [ ] Bug
- [x] Enhancement  
      POEM [#044](OpenMDAO/POEMs#100)

### Description

POEM 044 Proposes the addition of OpenMDAO-specific warning classes which can be more precisely controlled by the user.

### Example

N/A

### Environment

N/A

### PR DESCRIPTION
**Pull request by robfalck**
### Summary

Changes warnings issued by OpenMDAO to new OpenMDAO-specific warning classes.

| Warning Class               | Description                                                                                |
|-----------------------------|--------------------------------------------------------------------------------------------|
| CacheWarning                | A cache is invalid and must be discarded.                                                  |
| CaseRecorderWarning         | Warning encountered by a case recorder or case reader.                                     |
| DerivativesWarning          | Warning issued when the approximated partials or coloring cannot be evaluated as expected. |
| DriverWarning               | Warning encountered during driver execution.                                               |
| OMDeprecationWarning        | Warning of the use of deprecated OpenMDAO features.                                        |
| SetupWarning                | General errors encountered during setup.                                                   |
| SolverWarning               | Warning encountered during solver execution.                                               |
| UnusedOptionWarning         | A given option or argument has no effect.                                                  |

The following are subclasses of SetupWarning, allowing them all to be silenced by setting om.SetupWarning to ignore:

| Warning Class                 | Description                                                                                |
|-----------------------------|--------------------------------------------------------------------------------------------|
| DistributedComponentWarning | Warning specific to a distributed component.                                               |
| MPIWarning                  | Warning dealing with the availability of MPI.                                              |
| PromotionWarning            | Warning dealing with the promotion of an input or output.                                  |
| UnitsWarning                | Warning which is issued when unitless variable is connected to a variable with units.      |

### Related Issues

- Resolves #1944

### Backwards incompatibilities

None

### New Dependencies

None

### CODE REVIEW

File: openmdao/docs/features/warnings.rst [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+------------------------------------
+
+Class **OpenMDAOWarning** serves as the base-class for all OpenMDAO-specific warnings.
+All OpenMDAO-specific warnings default to a filter of 'always'.
+The following table shows all OpenMDAOWarning-derived classes.
+
++-----------------------------+--------------------------------------------------------------------------------------------+
+| Warning Class               | Description                                                                                |
++=============================+============================================+===============================================+
+| CacheWarning                | A cache is invalid and must be discarded.                                                  |
```
Comment (naylor-b) [inline openmdao/docs/features/warnings.rst:22]: Descriptions of these warnings should be made consistent, i.e., some start with "Warning (issued/encountered) (when/by) ..." and others don't.

File: openmdao/docs/features/warnings.rst [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+| DistributedComponentWarning | Warning specific to a distributed component.                                               |
+| MPIWarning                  | Warning dealing with the availability of MPI.                                              |
+| PromotionWarning            | Warning dealing with the promotion of an input or output.                                  |
+| UnitsWarning                | Warning which is issued when unitless variable is connected to a variable with units.      |
++-----------------------------+--------------------------------------------------------------------------------------------+
+
+Filtering Warnings
+------------------
+
+Python's built-in warning filtering system can be used to control which warning are displayed when using OpenMDAO.
```
Comment (naylor-b) [inline openmdao/docs/features/warnings.rst:50]: warning -> warnings

File: openmdao/solvers/linesearch/tests/test_backtracking.py [hunk 3]
```diff
@@ -443,6 +443,7 @@ def test_linesearch_vector_bound_enforcement(self):
             assert_near_equal(top['comp.z'][ind], [1.5], 1e-8)
 
         msg = (f"'comp.z' exceeds lower bounds\n  Val: [1.33333333 1.33333333 1.33333333]\n  Lower: [1.5 1.5 1.5]\n")
+        top.run_model()
```
Comment (naylor-b) [inline openmdao/solvers/linesearch/tests/test_backtracking.py:446]: Was this run_model call just leftover from debugging or was it intentional?

File: openmdao/test_suite/tests/test_warnings.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
+import unittest
+import openmdao.api as om
+
+class c(unittest.TestCase):
```
Comment (naylor-b) [inline openmdao/test_suite/tests/test_warnings.py:4]: class could use a more descriptive name

File: openmdao/warnings.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    Returns
+    -------
+    str
+        A string representation of a markdown table of the warning options.
+    """
+    s = io.StringIO()
+    max_name_len = max([len(_class.__name__) for _class in _warnings])
+    max_desc_len = max([len(' '.join(c.__doc__.split())) for c in _warnings])
+
+    name_header = "Option Name"
```
Comment (naylor-b) [inline openmdao/warnings.py:195]: Should this be 'Warning Class' instead of 'Option Name' ?

**Review by naylor-b — CHANGES_REQUESTED** [review]

File: openmdao/core/tests/test_group.py [hunk 6]
```diff
@@ -1975,10 +1975,7 @@ def configure(self):
 
         p = om.Problem(model=SimpleGroup())
 
-        with assert_warning(UserWarning,
```
[the remark this answers is not in the document]
Comment (robfalck) [reply openmdao/core/tests/test_group.py:1978]: We discussed this warning, which basically tells the user that if they promote '*' then src_indices only applies to inputs, and decided that it should be obvious.

**Review by Kenneth-T-Moore — APPROVED** [review]
This looks good and makes sense. I notice there aren't any Driver warnings, but looking through the code, there really isn't much to warn about that isn't covered by the derivatives warnings.

File: openmdao/test_suite/tests/test_warnings.py (hunk 4 above)
Comment (robfalck) [reply openmdao/test_suite/tests/test_warnings.py]: resolved

File: openmdao/warnings.py (hunk 5 above)
Comment (robfalck) [reply openmdao/warnings.py]: resolved

File: openmdao/solvers/linesearch/tests/test_backtracking.py (hunk 3 above)
Comment (robfalck) [reply openmdao/solvers/linesearch/tests/test_backtracking.py]: resolved

File: openmdao/docs/features/warnings.rst (hunk 2 above)
Comment (robfalck) [reply openmdao/docs/features/warnings.rst]: resolved

File: openmdao/docs/features/warnings.rst (hunk 1 above)
Comment (robfalck) [reply openmdao/docs/features/warnings.rst]: resolved

File: openmdao/docs/features/warnings.rst [hunk 7]
```diff
@@ -0,0 +1,93 @@
+.. _feature_warnings:
+
+***************
+Warning Control
+***************
+
+OpenMDAO has several classes of warnings that may be raised during operation.
+In general, these warnings are useful and the user should pay attention to them.
+Sometimes these warnings can be unnecessarily noisy.
+Filtering out noisy "low-priority" warning can make other more important ones more obvious.
```
[the remark this answers is not in the document]
Comment (robfalck) [reply openmdao/docs/features/warnings.rst]: resolved

File: openmdao/core/analysis_error.py [hunk 8]
```diff
@@ -1,8 +1,6 @@
 """
 OpenMDAO custom error: AnalysisError.
 """
-import inspect
-from openmdao.utils.general_utils import simple_warning
 from openmdao.utils.general_utils import _warn_simple_format, reset_warning_registry
```
[the remark this answers is not in the document]
Comment (robfalck) [reply openmdao/core/analysis_error.py]: resolved

**Review by naylor-b — APPROVED** [review]

File: openmdao/core/notebook_mode.py [hunk 9]
```diff
@@ -1,7 +1,7 @@
 """Checking for interactive notebook mode."""
 import sys
 
-from openmdao.utils.general_utils import simple_warning
+from openmdao.warnings import issue_warning
```
Comment (swryan) [inline openmdao/core/notebook_mode.py:4]: this file has been renamed to `notebook_utils.py` on `master`

**Review by swryan — CHANGES_REQUESTED** [review]

**Review by swryan — APPROVED** [review]

**Pull request merged**